receivers/jira/jira.go

```diff
@@ -0,0 +1,376 @@
+package jira
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/types"
+
+	"github.com/grafana/alerting/logging"
+	"github.com/grafana/alerting/receivers"
+	"github.com/grafana/alerting/templates"
+)
+
+const (
+	MaxSummaryLenRunes     = 255
+	MaxDescriptionLenRunes = 32767
+	adfDocOverhead         = 87
+)
+
+// Notifier implements a Notifier for JIRA notifications. Can use V2 and V3 API to create issues, depending on Config.URL
+// https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-issues/#api-rest-api-3-issue-post
+type Notifier struct {
+	*receivers.Base
+	tmpl    *templates.Template
+	log     logging.Logger
+	ns      receivers.WebhookSender
+	conf    Config
+	retrier *notify.Retrier
+}
+
+func New(cfg Config, meta receivers.Metadata, template *templates.Template, sender receivers.WebhookSender, logger logging.Logger) *Notifier {
+	return &Notifier{
+		Base:    receivers.NewBase(meta),
+		log:     logger,
+		ns:      sender,
+		tmpl:    template,
+		conf:    cfg,
+		retrier: &notify.Retrier{RetryCodes: []int{http.StatusTooManyRequests}},
+	}
+}
+
+// Notify implements the Notifier interface.
+func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	if n.conf.URL == nil {
+		// this should not happen but it's better to avoid panic
+		return false, fmt.Errorf("missing JIRA URL")
+	}
+	var (
+		alerts = types.Alerts(as...)
+	)
+
+	key, err := notify.ExtractGroupKey(ctx)
+	if err != nil {
+		return false, err
+	}
+	logger := n.log.New("group_key", key.Hash())
+	logger.Debug("executing Jira notification")
+
+	existingIssue, shouldRetry, err := n.searchExistingIssue(ctx, logger, key.Hash(), alerts.HasFiring())
+	if err != nil {
+		return shouldRetry, fmt.Errorf("failed to look up existing issues: %w", err)
+	}
+
+	var method string
+	var path string
+	if existingIssue == nil {
+		// Do not create new issues for resolved alerts
+		if alerts.Status() == model.AlertResolved {
+			return false, nil
+		}
+		path = "issue"
+		method = http.MethodPost
+		logger.Debug("create new issue")
+	} else {
+		path = "issue/" + url.PathEscape(existingIssue.Key)
+		method = http.MethodPut
+		logger.Debug("updating existing issue", "issue_key", existingIssue.Key)
+	}
```

Comment:
Just a suggestion, you can ignore the `method` and `path` changes, but please change "create" to "creating" in the debug log line.
```suggestion
	method := http.MethodPost
	path := "issue"
	if existingIssue == nil {
		// Do not create new issues for resolved alerts
		if alerts.Status() == model.AlertResolved {
			return false, nil
		}
		logger.Debug("creating new issue")
	} else {
		path = "issue/" + url.PathEscape(existingIssue.Key)
		method = http.MethodPut
		logger.Debug("updating existing issue", "issue_key", existingIssue.Key)
	}
```

Reply:
I did this intentionally. I think it more readable this way when you do actions in if branches. Anyway, I am fine with this change